website/blog/2024-08-06-langgraph-pyfunc/index.md

```diff
@@ -0,0 +1,236 @@
+
+To quickly summarize the code below, we look to create a class that inherits from `mlflow.pyfunc.PythonModel`. Within that class we create a `predict` method that is standard to MLflow. We also create a variety of helpers to facilitate chatbot functionality. One notable call-out is that in the `predict` method, we perform the import of our custom `compiled_graph` object from the `graph_chain` python module created in the above step. The reason that this is possible is due to the fact that we execute the stored script when loading the model into an inference environment, just as you would execute any python script. Running the script makes the `graph_chain` reference available to the REPL.
+
+After creating this class, we leverage the standard APIs to log the model. 
+
+```python
+import mlflow
+from mlflow.pyfunc import PythonModel
+from mlflow.models.utils import _convert_llm_input_data
+from langchain_core.messages import HumanMessage
+
+
+class LangGraphCustomChain(PythonModel):
+    messages: list = []
+    graph = None
+
+    def _format_input(self, text_input) -> dict:
+        """Convert a string to a user message."""
+        # Use an MLflow utility to clean the input
+        converted_input = _convert_llm_input_data(text_input)
+
+        # Handle lists (production logic would be more complex)
+        if isinstance(converted_input, list):
+            converted_input = next(iter(converted_input))
+
+        # Format as dict of messages and return
+        return {"messages": self.messages + [HumanMessage(converted_input)]}
+
+    def _parse_output(self, message_output: dict) -> str:
+        """Get the last message from the output."""
+        # Append to message history
+        self.messages = message_output["messages"]
+
+        # Return the last message str
+        return self.messages[-1].content
+
+    def load_context(self, context):
+        # Validate enviornment variable secrets are present
+        required_secrets = ["OPENAI_API_KEY", "LANGSMITH_API_KEY"]
+
+        if missing_keys := [key for key in required_secrets if not os.environ.get(key)]:
+            raise ValueError(f"The following keys are missing: {missing_keys}")
+
+        # Add project environent variables if not present
+        os.environ["LANCHAIN_TRACING_V2"] = (
+            "true"
+            if not os.environ.get("LANGCHAIN_TRACING_V2")
+            else os.environ.get("LANGCHAIN_TRACING_V2")
+        )
+        os.environ["LANGCHAIN_PROJECT"] = (
+            "LangGraph MLflow Tutorial"
+            if not os.environ.get("LANGCHAIN_TRACING_V2")
+            else os.environ.get("LANGCHAIN_TRACING_V2")
+        )
+
+    def predict(self, context, input_data: str, params: dict = None):
+        # Import from the `graph_chain.py` module specified in the `code_paths` argument
+        from graph_chain import compiled_graph
+
+        self.graph = compiled_graph
```

Comment:
Will this work if placed in the `load_context` scope? 
If yes, let's put it there to save on object creation references with each invocation call to predict. 
If no, we need to add a comment and explanation about why this doesn't work when defining the inference object.

Reply:
Great catch